### DISCUSSION TITLE
Smart Locks and unlocking

### DISCUSSION THREAD

**Discussion by illuzn** [make-all/tuya-local#2114] [General]
According to this document here: https://developer.tuya.com/en/docs/iot/ble-doorlock-developer-function-document?id=K9fwaai7m9wt3#title-35-Unlock%20with%20Bluetooth%20connections

Once paired, the Tuya server sets a key using `DPS 60` - this key is stored in flash. I cannot retrieve this code via debugging on Tuya IOT (not sure if this is deliberately hidden except to the Tuya app or I'm doing something wrong). It would seem this might be by design for security reasons.

Importantly, you need send this key via `DPS 61` to remotely unlock the device. 

It also seems `DPS64` called `password_offline_time` stores a UNIX UTC timestamp when the offline key was set. I've only had my locks installed for a day but presumably there is a rolling password validity.

This answers why remote unlocking is not available in #2085, #2059, #1885 (and probably many other issues).

Implementing unlocking appears to be outside the scope of tuya-local (unless somebody can figure out how DPS 60 works) because itotherwise  requires a cloud connection.

@make-all Feel free to add this to the readme (or link here) because no doubt there will continue to be repeated issues asking why remote unlock is not possible.

**Comment by make-all** [comment]
Yes, that is pretty much a summary.  If someone can work out the details, then passcode unlock support from HA can probably be added, but for now, the majority of locks are not supporting remote unlocking due to the complex nature of the interactions, and the documentation is technically there, but doesn't really explain the flow of how things should work or give examples.
